linera-witty/src/exported_function_interface/guest_interface.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Representation of the interface a guest Wasm instance uses for a host function.
+//!
+//! The [`GuestInterface`] trait is implemented for the tuple pair of the flattened representation
+//! of the host function's parameters type and results type. The trait then maps the flattened host
+//! function signature into the interface guest Wasm instances expect. This involves selecting
+//! where the host function's parameters and results are stored and if the guest Wasm instance
+//! needs to provide additional parameters to the function with the addresses to store the
+//! parameters and/or results in memory.
+//!
+//! The [canonical ABI][flattening] describes a `MAX_FLAT_PARAMS` limit on the number of flat types
+//! that can be sent to the function through Wasm parameters. If more flat parameters are needed,
+//! then all of them are stored in memory instead, and the guest Wasm instance must provide a
+//! single parameter with the address in memory of where the parameters are stored. The same must
+//! be done if more flat result types than the `MAX_FLAT_RESULTS` limit defined by the [canonical
+//! ABI][flattening] need to be returned from the function. In that case, the guest Wasm instance
+//! is responsible for sending an additional parameter with an address in memory where the results
+//! can be stored. See [`ResultStorage`] for more information.
+//!
+//! [flattening]: https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#flattening
+
+use super::result_storage::ResultStorage;
+use crate::{
+    memory_layout::FlatLayout, primitive_types::FlatType, util::Split, GuestPointer,
+    InstanceWithMemory, Layout, Memory, Runtime, RuntimeError, RuntimeMemory, WitLoad,
+};
+use frunk::HList;
+use std::ops::Add;
+
+/// Representation of the interface a guest Wasm instance uses for a host function.
+///
+/// See the module level documentation for more information.
+pub trait GuestInterface {
+    /// The flattened layout of the host function's parameters type.
+    ///
+    /// This is used as a generic input type in all implementations, but is needed here so it can
+    /// be used in the constraints of the [`Self::lift_parameters`] method.
+    type FlatHostParameters: FlatLayout;
+
+    /// The flat types the guest Wasm instances use as parameters for the host function.
+    type GuestParameters: FlatLayout;
```

Comment:
Done. Good question. It's quite subtle, because while there's a need to distinguish between non-flat and flat host parameters, the same can't be said about the guest parameters. But now I think this is clearer because in the `ExportedFunctionInterface` implementation there's the explicit `type GuestParameters = <Self::GuestInterface as GuestInterface>::FlatGuestParameters`, which should make the fact that they are the same more explicit :thinking: Thanks!